websockets-jsr/src/main/java/io/undertow/websockets/jsr/WebSocketSessionRemoteEndpoint.java

```diff
@@ -268,57 +268,78 @@ public void sendBinary(final ByteBuffer data) throws IOException {
             data.clear(); //for some reason the TCK expects this, might as well just match the RI behaviour
         }
 
-        @Override
+       @Override
         public void sendText(final String partialMessage, final boolean isLast) throws IOException {
-            if(partialMessage == null) {
+            if (partialMessage == null) {
                 throw JsrWebSocketMessages.MESSAGES.messageInNull();
             }
-            if (binaryFrameSender != null) {
-                throw JsrWebSocketMessages.MESSAGES.cannotSendInMiddleOfFragmentedMessage();
-            }
-            if (textFrameSender == null) {
-                textFrameSender = undertowSession.getWebSocketChannel().send(WebSocketFrameType.TEXT);
-            }
+
+            StreamSinkFrameChannel sender = getTextFrameSender();
+
             try {
-                Channels.writeBlocking(textFrameSender, WebSocketUtils.fromUtf8String(partialMessage));
-                if(isLast) {
-                    textFrameSender.shutdownWrites();
+                Channels.writeBlocking(sender, WebSocketUtils.fromUtf8String(partialMessage));
+                if (isLast) {
+                    sender.shutdownWrites();
                 }
-                Channels.flushBlocking(textFrameSender);
+                Channels.flushBlocking(sender);
             } finally {
                 if (isLast) {
-                    textFrameSender = null;
+                    clearTextFrameSender();
                 }
             }
-
         }
 
         @Override
         public void sendBinary(final ByteBuffer partialByte, final boolean isLast) throws IOException {
-
             if(partialByte == null) {
                 throw JsrWebSocketMessages.MESSAGES.messageInNull();
             }
-            if (textFrameSender != null) {
-                throw JsrWebSocketMessages.MESSAGES.cannotSendInMiddleOfFragmentedMessage();
-            }
-            if (binaryFrameSender == null) {
-                binaryFrameSender = undertowSession.getWebSocketChannel().send(WebSocketFrameType.BINARY);
-            }
+
+            StreamSinkFrameChannel sender = getBinaryFrameSender();
```

Comment:
Im going to talk about this with @ropalka . But it seems its technically possible for competing send to undercut one another, though I might be missing how this piece is being used.

Reply:
I would like to share my understanding and seek your feedback.

For example, in the case of sending a **Text** message, I assume that a single message is divided into multiple frames for transmission. In this process, the `TextFrameSender` class variable is used to call `sendText(partialMessage, isLast)` multiple times, sending partial messages through the same channel. (During this process, other `send()` methods are not allowed to send messages.)

However, if partial frames are being sent from multiple threads, I believe there is a possibility that one thread might set `textFrameSender` to null first, and then another thread could call `Channels.flushBlocking(StreamSinkFrameChannel)`, leading to a NullPointerException (NPE).

To address this, I have modified the code to ensure synchronization when accessing the shared resource, the `textFrameSender` class variable.

Additionally, could you please clarify what specific scenarios you had in mind regarding `'competing sends interfering with one another'`? 
Understanding this will help me consider those cases and make further improvements accordingly.